src/module.c

```diff
@@ -3048,6 +3087,394 @@ int RM_HashGet(RedisModuleKey *key, int flags, ...) {
+ * reclaims its memory.
+ *
+ * Returns REDISMODULE_OK on success. On failure, REDISMODULE_ERR is returned
+ * and `errno` is set as follows:
+ *
+ * - EINVAL if called with a NULL key
+ * - ENOTSUP if the key refers to a value of a type other than stream or if the
+ *   key is empty
+ * - EBADF if the key was not opened for writing or if no stream iterator is
+ *   associated with the key
+ */
+int RM_StreamIteratorStop(RedisModuleKey *key) {
+    if (!key) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    } else if (!key->value || key->value->type != OBJ_STREAM) {
+        errno = ENOTSUP;
+        return REDISMODULE_ERR;
+    } else if (!key->iter) {
+        errno = EBADF;
+        return REDISMODULE_ERR;
+    }
+    zfree(key->iter);
+    key->iter = NULL;
+    return REDISMODULE_OK;
+}
+
+/* Finds the next stream entry and returns its stream ID and the number of
+ * fields.
+ *
+ * - `key`: Key for which a stream iterator has been started using
+ *   RedisModule_StreamIteratorStart().
+ * - `id`: The stream ID returned. NULL if you don't care.
+ * - `numfields`: The number of fields in the found stream entry. NULL if you
+ *   don't care.
+ *
+ * Returns REDISMODULE_OK and sets `*id` and `*numfields` if an entry was found.
+ * If there are no more entries, REDISMODULE_EOF is returned. On failure,
+ * REDISMODULE_ERR is returned and `errno` is set as follows:
+ *
+ * - EINVAL if called with a NULL key
+ * - ENOTSUP if the key refers to a value of a type other than stream or if the
+ *   key is empty
+ * - EBADF if no stream iterator is associated with the key
+ *
+ * Use RedisModule_StreamIteratorNextField() to retrieve the fields and values.
+ * See the example at RedisModule_StreamIteratorStart().
+ */
+int RM_StreamIteratorNextID(RedisModuleKey *key, RedisModuleStreamID *id, long *numfields) {
+    if (!key) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    } else if (!key->value || key->value->type != OBJ_STREAM) {
+        errno = ENOTSUP;
+        return REDISMODULE_ERR;
+    } else if (!key->iter) {
+        errno = EBADF;
+        return REDISMODULE_ERR;
+    }
+    streamIterator *si = key->iter;
+    int64_t num;
+    streamID streamid;
+    if (streamIteratorGetID(si, &streamid, &num)) {
+        if (id) {
+            id->ms = streamid.ms;
+            id->seq = streamid.seq;
+        }
+        if (numfields) *numfields = num;
+        return REDISMODULE_OK;
+    } else {
+        /* No entry found. */
+        if (id) {
+            id->ms = 0;
+            id->seq = 0;
+        }
+        if (numfields) *numfields = 0;
+        return REDISMODULE_EOF;
+    }
+}
+
+/* Retrieves the next field of the current stream ID and its corresponding value
+ * in a stream iteration. This function should be called directly after calling
+ * RedisModule_StreamIteratorNextID() and not more than `numfields` times. If
+ * called more times than the number of fields, the behaviour is undefined.
```

Comment:
i'm not comfortable with this undefined behavior.
i think we should check `si->lp_ele` (maybe add an `streamIteratorIsValidField`, or modify `streamIteratorGetField` to return an error), and handle that too (possibly setting errno to ENOENT).

Reply:
I'll try this. I was planning to store a field counter of the current entry in the key struct and use it to return an error in NextField(), but if it can be done in streamIterator itself, it's better of course.

Reply:
Solved. I followed my original idea and stored a counter in the key struct.

Reply:
i'm curious, did you try the `si->lp_ele` idea and found some problem with it, or just didn't want to do any changes in t_stream.c ?

Reply:
The latter. I haven't spent enough time to understand the internals of t_stream.c and with the union in the key struct, it doesn't really cost anything extra to put it there... Do you prefer the `si->lp_ele` solution?

Reply:
i prefer it since it seems safer (won't cause NPD in case of a mismatch between the count and what's actually in the listpack, see #7807). 
I suppose such a change would also have such a side effect benefit for other pieces of in stream that use this function (even if they do run a loop that exit only on field count), but i suppose for modules this safety is of higher concern.

from a quick look i took now, in order to make it safe, `streamIteratorGetField` should exit when `si->lp_ele` is NULL.
maybe for extra safety, it would also be a nice idea to nullify it in two places in `streamIteratorGetID` when it returns 0 due to running out of range.

Reply:
I'm looking at this.

* `streamIteratorGetField` returns `void`. It would need to return `int` to indicate if there was a field or not.
* I can't see the two places where you want to nullify `si->lp_ele`. lpNext() and lpPrev() already return NULL when running out of range. Maybe you can enlighten me?

I suppose this can also be changed as a separate improvement after merging this PR. I don't feel very confident about the listpack and rax structure.

Reply:
Would the issue in `streamIteratorGetField()` be that if the first `lpNext()` returns NULL since it is running out of range, the second `lpNext()` will get NULL in its listpack pointer, which triggers an assert in `lpNext()`.
I guess `streamIteratorGetField()` needs to bail out after attempting to get the missing fieldptr, and then it might need to nullify the returned field and value ptrs before returning. This can later be checked in `RM_StreamIteratorNextField()` to make sure that more fields existed.

Reply:
the two places i meant that i think `streamIteratorGetID` should nullify `si->lp_ele` are:
https://github.com/redis/redis/blob/9e56d3969a3d19ef73d7042f91ba34bef39ed189/src/t_stream.c#L1121
https://github.com/redis/redis/blob/9e56d3969a3d19ef73d7042f91ba34bef39ed189/src/t_stream.c#L1132

i think that if we do that, and change `streamIteratorGetField` to return an error when `si->lp_ele` is NULL on entry, should make all of this safe.

i'm ok leaving this issue for some future day, and also ok to change the PR to use that instead of counting the fields.
up to you if you wanna try it.

Reply:
OK, thanks. I tried it, but GetField() doens't reach `si->lp_ele == NULL` after the last field in an entry, so it just continues returning fields from the next stream entry. My guess that we need a numfieldsleft counter for this, either inte the streamIterator or in the API...

Btw, this PR reached :100: comments. :)

Reply:
ohh, right. for some reason i thought the next entry is in the next rax node.
sorry.